manticore/core/memory.py

```diff
@@ -1045,6 +1085,74 @@ def _try_get_solutions(self, address, size, access, max_solutions=0x1000, force=
         return solutions
 
 
+class LazySMemory(SMemory):
+    '''
+    A fully symbolic memory.
+
+    Currently does not support cross-page reads/writes.
+    '''
+    def __init__(self, constraints, *args, **kwargs):
+        super(LazySMemory, self).__init__(constraints, *args, **kwargs)
+
+    def mmap(self, addr, size, perms, name=None, **kwargs):
+        assert addr is None or isinstance(addr, (int, long)), 'Address must be concrete'
+        assert addr < self.memory_size, 'Address too big'
+        assert size > 0
+
+        if addr is not None:
+            addr = self._floor(addr)
+
+        size = self._ceil(size)
+
+        addr = self._search(size, addr)
+
+        # It should not be allocated
+        for i in xrange(self._page(addr), self._page(addr + size)):
+            assert i not in self._page2map, 'Map already used'
+
+        m = ArrayMap(addr, size, perms, self.memory_bit_size, name=name)
+
+        self._add(m)
+
+        logger.debug('New symbolic memory map @{:x} size:{:x}'.format(addr, size))
+
+        return addr
+
+    def _deref_can_succeed(self, map, address, size):
+        if not issymbolic(address):
+            return address >= map.start and address + size < map.end
+        else:
+            constraint = Operators.AND(address >= map.start, address + size < map.end)
+            return solver.can_be_true(self.constraints, constraint)
+
+    def map_containing(self, address):
+        if not issymbolic(address):
+            return super(LazySMemory, self).map_containing(address)
+        else:
+            found = None
+            for m in self._maps:
+                if self._deref_can_succeed(m, address, 1):
+                    if found:
+                        raise InvalidMemoryAccess(address, 'r')
+                    found = m
+            return found
+
+    def read(self, address, size, force=False):
+        m = self.map_containing(address)
+        if isinstance(m, ArrayMap):
```

Comment:
could we simplify this by assuming all maps in the LazySMemory will be ArrayMaps?

Reply:
I am using `FileMap`s for things loaded by the elf loader; we should probably back them with initialized `ArrayMap`s

Reply:
Should not you upgrade the map to symbolic when writting a symbolic value?